### PR TITLE
Update support for cuda reductions to work with types < 4 bytes in size

### LINKED ISSUE #4219: Update support for cuda reductions to work with types < 4 bytes in size
Related to #3531

### PR DESCRIPTION
**Pull request by e10harvey**
Related to #4101. Merging this PR will merge #4101.

Fixes #4219.

### CODE REVIEW

**Comment by e10harvey** [comment]
retest this please

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 1]
```diff
@@ -1115,13 +1122,17 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
 
       KOKKOS_ASSERT(block_size > 0);
 
-      m_scratch_space = cuda_internal_scratch_space(
+      // TODO: down casting these uses more space than required?
+      m_scratch_space = (size_type*)cuda_internal_scratch_space(
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1126]: static_cast instead

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 2]
```diff
@@ -1115,13 +1122,17 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
           m_policy.space(), ValueTraits::value_size(ReducerConditional::select(
                                 m_functor, m_reducer)) *
                                 block_size /* block_size == max block_count */);
+
+      // Intentionally do not downcast to m_scratch_flags since we use Cuda
+      // atomics in Kokkos_Cuda_ReduceScan.hpp
       m_scratch_flags =
           cuda_internal_scratch_flags(m_policy.space(), sizeof(size_type));
-      m_unified_space = cuda_internal_scratch_unified(
+      m_unified_space = (size_type*)cuda_internal_scratch_unified(
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1135]: static cast instead

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 3]
```diff
@@ -212,6 +225,70 @@ class half_t {
...
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  void atomic_mul(const volatile half_t src) const volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    alignas(4) float lhs_tmp = __half2float(val),
+                     rhs_tmp = __half2float(src.val);
+    volatile float* lhs      = &lhs_tmp;
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:260]: instead:
```
volatile float& lhs = lhs_tmp;
```
avoids the array syntax.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 4]
```diff
@@ -212,6 +225,70 @@ class half_t {
   KOKKOS_FUNCTION
   half_t(unsigned long long rhs) : val(cast_to_half(rhs).val) {}
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void atomic_add(const volatile half_t src) const volatile {
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:230]: either copy the content to where you use it (it looks like its only called once?) or rename to something like impl_volatile_add.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 5]
```diff
@@ -212,6 +225,70 @@ class half_t {
...
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    const impl_type new_val =
+        __float2half(__half2float(const_cast<impl_type&>(val)) +
+                     __half2float(const_cast<const impl_type&>(src.val)));
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  void atomic_mul(const volatile half_t src) const volatile {
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:255]: either copy the content to where you use it (it looks like its only called once?) or rename to something like impl_volatile_mult.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 6]
```diff
@@ -212,6 +225,70 @@ class half_t {
...
+    // *lhs = __half2float(val);
+    // *rhs = __half2float(src.val);
+    // lhs[0] *= rhs[0];
+    // val = __float2half(*lhs);
+
+    // 16 bit arithmetic on device follow by 16bit atomic store.
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    const impl_type new_val =
+        const_cast<impl_type&>(val) * const_cast<const impl_type&>(src.val);
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:280]: maybe clal explicitly the CUDA atomic store?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 7]
```diff
@@ -290,7 +367,19 @@ class half_t {
...
-    val = cast_to_half(rhs).val;
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    impl_type new_val = cast_to_half(rhs).val;
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    alignas(4) float rhs_tmp = __half2float(new_val);
+    volatile float* new_lhs  = &rhs_tmp;
+    val                      = __float2half(new_lhs[0]);
+#else
+    Kokkos::atomic_store(val_ptr, *((uint16_t*)&new_val));
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:378]: call explicitly the CUDA atomic store/volatile assignment?

File: core/src/Kokkos_Parallel_Reduce.hpp [hunk 8]
```diff
@@ -91,8 +92,8 @@ struct Sum {
   KOKKOS_INLINE_FUNCTION
   void join(value_type& dest, const value_type& src) const { dest += src; }
 
-  KOKKOS_INLINE_FUNCTION
-  void join(volatile value_type& dest, const volatile value_type& src) const {
+  KOKKOS_INLINE_FUNCTION void join(volatile value_type& dest,
```
Comment (crtrott) [inline core/src/Kokkos_Parallel_Reduce.hpp:95]: undo this change

File: core/src/Kokkos_Parallel_Reduce.hpp [hunk 9]
```diff
@@ -136,8 +137,8 @@ struct Prod {
   KOKKOS_INLINE_FUNCTION
   void join(value_type& dest, const value_type& src) const { dest *= src; }
 
-  KOKKOS_INLINE_FUNCTION
-  void join(volatile value_type& dest, const volatile value_type& src) const {
+  KOKKOS_INLINE_FUNCTION void join(volatile value_type& dest,
```
Comment (crtrott) [inline core/src/Kokkos_Parallel_Reduce.hpp:140]: undo this change

File: core/src/impl/Kokkos_FunctorAdapter.hpp [hunk 10]
```diff
@@ -1334,9 +1334,9 @@ struct FunctorValueTraits<FunctorType, ArgTag,
   using functor_type = FunctorType;
 
   static_assert(
-      IS_VOID || IS_REJECT || 0 == (sizeof(ValueType) % sizeof(int)),
+      IS_VOID || IS_REJECT || sizeof(ValueType) != 3,
```
Comment (crtrott) [inline core/src/impl/Kokkos_FunctorAdapter.hpp:1337]: do you actually need that? I think your logic would just use a value_count of 3 and a size_type of int8_t? Might be worthwhile to check with a struct of 3 chars and alignas(3)

File: core/unit_test/TestHalfOperators.hpp [hunk 11]
```diff
@@ -269,6 +269,88 @@ enum OP_TESTS {
...
+    actual_lhs(GT)   = h_lhs > h_rhs;
+    expected_lhs(GT) = d_lhs > d_rhs;
+
+    actual_lhs(GE)   = h_lhs >= h_rhs;
+    expected_lhs(GE) = d_lhs >= d_rhs;
+
+    actual_lhs(EQ)   = h_lhs == h_rhs;
+    expected_lhs(EQ) = d_lhs == d_rhs;
+
+    // cast away volatile on lhs to supress:
```
Comment (crtrott) [inline core/unit_test/TestHalfOperators.hpp:325]: spelling: suppress

**Review by crtrott — CHANGES_REQUESTED** [review]
Largely I think this is good. Main change is to not add "public" member functions atomic_add to half_t. Make them private ones and also don't name them "atomic_add" since they are not that. See my comment.

**Comment by crtrott** [comment]
if we have this (with the suggested changes) there is no need for #4101 or?

**Comment by crtrott** [comment]
My review was done together with Damien.

**Comment by e10harvey** [comment]
> if we have this (with the suggested changes) there is no need for #4101 or?

Correct but we will still need the changes from #4101.

**Comment by e10harvey** [comment]
retest this please

**Comment by e10harvey** [comment]
Note sure what's going on with the jenkins-ci; I do not see any errors or failures in the console.

**Comment by e10harvey** [comment]
retest this please

**Comment by e10harvey** [comment]
I will push a fix for the CI failure shortly.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 7 above)
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp]: cuda does not have atomic store. __half supports volatile assignment but not loads.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 6 above)
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp]: can you point me to the CUDA atomic store in the programming guide?

File: core/src/impl/Kokkos_FunctorAdapter.hpp (hunk 10 above)
Comment (e10harvey) [reply core/src/impl/Kokkos_FunctorAdapter.hpp:1337]: i will add a test for this.

**Comment by e10harvey** [comment]
@crtrott, @Rombur, @dalg24, @masterleinad: Thank you for your input. This is ready for another round of reviews.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 12]
```diff
@@ -683,6 +746,68 @@ class half_t {
...
+        reinterpret_cast<const volatile uint16_t&>(lhs.val) !=
+        reinterpret_cast<const volatile uint16_t&>(rhs.val));
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator<(const volatile half_t& lhs,
+                        const volatile half_t& rhs) {
+    return static_cast<bool>(
+        reinterpret_cast<const volatile uint16_t&>(lhs.val) <
+        reinterpret_cast<const volatile uint16_t&>(rhs.val));
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:771]: uhm this seems rather wrong. Does this actually hold true? I mean reinterpret cast the bits as uint16_t and then compare? Isn't like the sign bit set for negative?

**Review by crtrott — CHANGES_REQUESTED** [review]
The volatile comparison seems rather wrong or?

**Comment by e10harvey** [comment]
> The volatile comparison seems rather wrong or?

I suppose we could cast to float here.

**Review by crtrott — APPROVED** [review]
I approve for now, we need a follow on for the impl half functions for volatile_add and volatile_mul. Did some testing of my own and the reductions seem to work also for custom types.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 13]
```diff
@@ -683,6 +746,92 @@ class half_t {
     return __half2float(val) >= __half2float(rhs.val);
 #endif
   }
+
+  KOKKOS_FUNCTION
+  friend bool operator==(const volatile half_t& lhs,
+                         const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<const impl_type&>(lhs.val) ==
+                             const_cast<const impl_type&>(rhs.val));
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:755]: I believe this won't provide the semantics that the signature promises, and worse results in undefined behavior:
"referring to a volatile object through a non-volatile glvalue results in undefined behavior. " https://en.cppreference.com/w/cpp/language/const_cast

Rather, the body of the `__CUDA_ARCH__` branch of this and later functions could be written as

```c++
impl_type lv = lhs.val;
impl_type rv = rhs.val;
return lv == rv;
```

That way the member accesses are actually volatile

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 14]
```diff
@@ -943,6 +1092,25 @@ KOKKOS_INLINE_FUNCTION
...
+template <>
+struct reduction_identity<Kokkos::Experimental::half_t> {
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static float sum() noexcept {
+    return 0.0F;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static float prod() noexcept {
+    return 1.0F;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static float max() noexcept {
+    return -65504.0F;
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:1107]: Does CUDA have `std::numeric_limits<__half>::max()/min()` available?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 15]
```diff
@@ -943,6 +1092,25 @@ KOKKOS_INLINE_FUNCTION
 }
 #endif
 }  // namespace Experimental
+
+// use float as the return type for sum and prod since cuda_fp16.h
+// has no constexpr functions for casting to __half
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:1097]: This comment raises an issue that should be addressed in the compiler and standard library - I'll call Max Katz's attention to it.

Comment (maxpkatz) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:1097]: Acknowledged and requested for cuda_fp16.h.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 16]
```diff
@@ -127,17 +131,111 @@ KOKKOS_INLINE_FUNCTION
...
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void impl_volatile_add(const volatile half_t src) const volatile {
+    // Kokkos::atomic_store and atomicAdd not available on cuda < 10
+#if (CUDA_VERSION < 10000)
+    alignas(4) volatile float lhs;
+    alignas(4) volatile float rhs;
+    lhs                         = __half2float(const_cast<impl_type&>(val));
+    rhs                         = __half2float(const_cast<impl_type&>(src.val));
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:149]: As noted elsewhere, this is negating the `volatile` semantics of the signature, and possibly UB. Load from the `volatile` member into a temporary, operate on that as needed, and then (if possible, since it's implied that it's not) store back to the member without casting away `volatile`.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 17]
```diff
@@ -299,30 +410,18 @@ class half_t {
 #ifdef __CUDA_ARCH__
     val += rhs.val;
 #else
-    val = __float2half(__half2float(val) + __half2float(rhs.val));
+    val     = __float2half(__half2float(val) + __half2float(rhs.val));
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:413]: Stray whitespace change

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 18]
```diff
@@ -299,30 +410,18 @@ class half_t {
...
-    val = const_cast<impl_type&>(val) + rhs.val;
-#else
-    // Use non-volatile val_ref to suppress:
-    // "warning: implicit dereference will not access object of type ‘volatile
-    // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) +
-                           __half2float(rhs.val));
-#endif
+  volatile half_t operator+=(const volatile half_t rhs) volatile {
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:419]: I believe the argument should be `const volatile half_t &rhs` (reference, not value) if you want to actually pass it through as `volatile`. Or, if you're ok with the compiler making the coherent read and copy to a temporary in the function call, use the original signature, `volatile half_t& operator+=(half_t rhs) volatile`. I think the latter is preferable, since there's no need to pass the argument around as a reference if we're guaranteed to need a temporary copy anyway.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 19]
```diff
@@ -350,26 +449,14 @@ class half_t {
 #ifdef __CUDA_ARCH__
     val -= rhs.val;
 #else
-    val          = __float2half(__half2float(val) - __half2float(rhs.val));
+    val     = __float2half(__half2float(val) - __half2float(rhs.val));
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:452]: Stray whitespace

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 20]
```diff
@@ -350,26 +449,14 @@ class half_t {
...
-    val = const_cast<impl_type&>(val) - rhs.val;
-#else
-    // Use non-volatile val_ref to suppress:
-    // "warning: implicit dereference will not access object of type ‘volatile
-    // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) -
-                           __half2float(rhs.val));
-#endif
+  volatile half_t operator-=(const volatile half_t rhs) volatile {
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:458]: Ditto from `operator+=`

File: core/unit_test/TestHalfOperators.hpp [hunk 21]
```diff
@@ -840,8 +919,34 @@ void __test_half_operators(half_t h_lhs, half_t h_rhs) {
...
+      ASSERT_NEAR(f_host.actual_lhs(op_test), f_host.expected_lhs(op_test),
+                  epsilon);
+    }
+  }
+
+  // is_trivially_copyable is false with the addition of explicit
+  // copy constructors that are required for supporting reductions
+  // ASSERT_TRUE(std::is_trivially_copyable<half_t>::value);
+
   // Check whether half_t is trivially copyable
```
Comment (PhilMiller) [inline core/unit_test/TestHalfOperators.hpp:949]: This comment line should be removed too

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 14 above)
Comment (PhilMiller) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:1107]: @maxpkatz this, too

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 16 above)
Comment (PhilMiller) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:149]: If, as an earlier comment suggests, CUDA doesn't actually support loads from `volatile half_t`, then do it with a `volatile` `memcpy`-like operation on `uint16_t` or the two underlying bytes

**Review by PhilMiller — CHANGES_REQUESTED** [review]
The requirement for `volatile` operations in the reduction infrastructure is (currently, until I'm successful with #4077) not just a syntactic/type lark, it's a semantic requirement for the compiler to generate consistently correct code. The implementation needs to consistently respect the spirit of those qualifiers, even if the compiler makes it hard to do the right thing.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 7 above)
Comment (PhilMiller) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp]: Is this what you're referring to?

```c++
#include <cuda_fp16.h>

int main(int argc, char **argv) {
  __half h = 0.0;
  volatile __half vh = 1.0;
  h = *(&vh);
  *(&vh) = h;

  return 0;
}
```

```
$ nvcc -c nvcc_half_t_test.cc 
nvcc_half_t_test.cc: In function ‘int main(int, char**)’:
nvcc_half_t_test.cc:7:10: warning: implicit dereference will not access object of type ‘volatile __half’ in statement
   *(&vh) = h;
   ~~~~~~~^~~
```

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 14 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:1107]: We have `Kokkos::Experimental::finite_min<float>::value` to replace std::numeric_limits in device code (but nothing for `__half`), also see https://stackoverflow.com/questions/24498140/is-there-a-cuda-equivalent-to-stdnumeric-limits.

Comment (maxpkatz) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:1107]: I will submit a request for `std::numeric_limits<__half>`. As indicated in the above SO post, CUDA doesn't provide numeric_limits in device code in general, however because these are constexpr functions, `--expt-relaxed-constexpr` generally works for providing this, so if we had a host version of this available it would likely be able to work in device code.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 22]
```diff
@@ -885,9 +885,14 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
   using value_type     = typename ValueTraits::value_type;
   using reference_type = typename ValueTraits::reference_type;
   using functor_type   = FunctorType;
-  using size_type      = Kokkos::Cuda::size_type;
-  using index_type     = typename Policy::index_type;
-  using reducer_type   = ReducerType;
+  // Conditionally set size_type to int16_t or int8_t if value_type is less than
+  // int32_t (Kokkos::Cuda::size_type)
+  using size_type = typename std::conditional<
```
[the remark this answers is not in the document]
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:890]: Good question. The name `size_type` was confusing to me as well. Here, `size_type` is <= the size of `value_type`. `size_type` is used to determine the word count, shared memory buffer size, and global memory buffer size before the reduction is performed. Within the reduction, atomics (other than implicit atomic stores and loads?) are only used for keeping track of metadata in `m_scratch_flags` which is why that local was left as type `Cuda::size_type*`. Also within the reduction, the word count is recomputed based on `size_type` and when calculating indexes into the shared/global memory buffers for performing the reduction, `size_type` is used again. So, without this compile-time conditional for setting `size_type`, our word_count and/or indexes are not correct for scalars < 4 bytes in size. For scalars > 4 bytes in size, indexing into shared/global memory relies on the block and grid dimensions to ensure that we index at the correct offset rather than at every 4 byte word; such that, when the join is performed, we have the correct data that was copied over in chunks of 4 bytes.

Christian suggested that I try out the approach you've used in #4077 for handling volatile reductions with `half_t` so I'll flag you in that PR after this merges.

[the remark this answers is not in the document]
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:890]: @PhilMiller: Does`word_size_type` make more sense?

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 13 above)
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:755]: Agreed. This seems so simple now... 🤦

Thank you!!

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (PhilMiller) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:890]: Yeah, `word_size_type` was the replacement I had in mind too.

I would support making that replacement in this PR, but support from @crtrott would probably mean more

[the remark this answers is not in the document]
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:890]: @crtrott: All the PR feedback changes requested from here and #4101 are sitting in my local branch. Do you want these pushed here or in a follow on?

**Review by PhilMiller — APPROVED** [review]
